### PR TITLE
refactor: Add unit testing and refactor code to follow Python standards.

### PR DESCRIPTION
**Pull request by bvanelli**
Hello @cybcon , I started writing this code on the last Hackergarten, but did not had the time to push it yet.

You wrote the integration test with the other part of the code. This tests how the applications work together, but it's too broad to test small functions inside the code. This PR adds the basic structure for unit testing too.

I also included other changes, here is the full changelog:

- Reformat file with community standards black, ruff and isort, with basic config. Black and ruff are formatters that make sure the code follows a stricter coding style. isort on the other hand makes sure the imports are sorted correctly.
- Format all variable namings using [PEP8](https://peps.python.org/pep-0008/). Here, I mostly took the suggestions from Pycharm, and some of it was done directly by black + ruff.
- Add two example unit tests for the part of the code that is currently testable.
- Add testing on CI using github actions and pytest. This makes sure the test would run on pull requests and on main branch.
- Improve code clarity by using f-strings, available since python 3.6. This should keep the complexity of writing custom logs much lower and improve readability.

I had to change some of the logic to make sure the code was testable by reducing the functions complexity, so some functions might be slightly different.

Let me know what you think and what you suggest.

### CODE REVIEW

File: src/app/modbus_server.py [hunk 1]
```diff
@@ -6,272 +6,320 @@
...
+        default=default_config_file,
+    )
+
+    args = parser.parse_args()
+    if "CONFIG_FILE" in os.environ:
+        config_file = os.environ["CONFIG_FILE"]
+    else:  # will either use the command line argument or the default value
+        config_file = args.config_file
+    # check if file actually exists
+    if not pathlib.Path(config_file).is_file():
```
Comment (cybcon) [inline src/app/modbus_server.py:263]: Why using pathlib.Path().is_file() instead of os.path.isfile()?

**Review by cybcon — CHANGES_REQUESTED** [review]

**Comment by cybcon** [comment]
Hi @bvanelli 
Thank you for this great contribution!

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      - uses: actions/checkout@v3
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install ruff pytest
+          if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
```
[the remark this answers is not in the document]
Comment (bvanelli) [reply .github/workflows/test.yaml]: Github actions uses ubuntu which should install without any problems. I added a line to change folders before running tests.

I also applied all your suggestions, maybe now the pipeline will go through.

**Review by cybcon — CHANGES_REQUESTED** [review]
Python test pipeline still fail.
In step pytest we get following error:
```
============================= test session starts ==============================
platform linux -- Python 3.9.18, pytest-8.0.0, pluggy-1.4.0
rootdir: /home/runner/work/modbus-server/modbus-server
collected 0 items / 2 errors

==================================== ERRORS ====================================
____________________ ERROR collecting tests/test_server.py _____________________
ImportError while importing test module '/home/runner/work/modbus-server/modbus-server/tests/test_server.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_server.py:2: in <module>
    from src.app.modbus_server import prepare_register
src/app/modbus_server.py:17: in <module>
    from pymodbus.datastore import (
E   ModuleNotFoundError: No module named 'pymodbus'
_____________________ ERROR collecting tests/test_utils.py _____________________
ImportError while importing test module '/home/runner/work/modbus-server/modbus-server/tests/test_utils.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_utils.py:2: in <module>
    from src.app.modbus_server import get_ip_address
src/app/modbus_server.py:17: in <module>
    from pymodbus.datastore import (
E   ModuleNotFoundError: No module named 'pymodbus'
=========================== short test summary info ============================
ERROR tests/test_server.py
ERROR tests/test_utils.py
!!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!!
============================== 2 errors in 0.10s ===============================
```

**Comment by bvanelli** [comment]
> Python test pipeline still fail. In step pytest we get following error:
> 
> ```
> ============================= test session starts ==============================
> platform linux -- Python 3.9.18, pytest-8.0.0, pluggy-1.4.0
> rootdir: /home/runner/work/modbus-server/modbus-server
> collected 0 items / 2 errors
> 
> ==================================== ERRORS ====================================
> ____________________ ERROR collecting tests/test_server.py _____________________
> ImportError while importing test module '/home/runner/work/modbus-server/modbus-server/tests/test_server.py'.
> Hint: make sure your test modules/packages have valid Python names.
> Traceback:
> /opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/importlib/__init__.py:127: in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
> tests/test_server.py:2: in <module>
>     from src.app.modbus_server import prepare_register
> src/app/modbus_server.py:17: in <module>
>     from pymodbus.datastore import (
> E   ModuleNotFoundError: No module named 'pymodbus'
> _____________________ ERROR collecting tests/test_utils.py _____________________
> ImportError while importing test module '/home/runner/work/modbus-server/modbus-server/tests/test_utils.py'.
> Hint: make sure your test modules/packages have valid Python names.
> Traceback:
> /opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/importlib/__init__.py:127: in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
> tests/test_utils.py:2: in <module>
>     from src.app.modbus_server import get_ip_address
> src/app/modbus_server.py:17: in <module>
>     from pymodbus.datastore import (
> E   ModuleNotFoundError: No module named 'pymodbus'
> =========================== short test summary info ============================
> ERROR tests/test_server.py
> ERROR tests/test_utils.py
> !!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!!
> ============================== 2 errors in 0.10s ===============================
> ```

I had the conditional wrong, could you try again?

**Review by cybcon — APPROVED** [review]

**Pull request merged**